sdk/containerregistry/container-registry/src/utils/retriableReadableStream.ts

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { AbortError } from "@azure/abort-controller";
+import { TransferProgressEvent } from "@azure/core-rest-pipeline";
+import { Readable } from "stream";
+
+export type ReadableStreamGetter = (offset: number) => Promise<NodeJS.ReadableStream>;
+
+export interface RetriableReadableStreamOptions {
+  /**
+   * Max retry count (greater than or equal to 0), undefined or invalid value means no retry
+   */
+  maxRetryRequests?: number;
+
+  /**
+   * Read progress event handler
+   */
+  onProgress?: (progress: TransferProgressEvent) => void;
+
+  /**
+   * On data event handler
+   */
+  onData?: (data: Buffer) => void;
+
+  /**
+   * On end (success) event handler
+   */
+  onEnd?: () => void;
+
+  /**
+   * Debug purpose only. Used to inject an unexpected end to existing internal stream,
+   * to test stream retry works well or not.
+   *
+   * When assign it to true, for next incoming "data" event of internal stream,
+   * RetriableReadableStream will try to emit an "end" event to existing internal
+   * stream to force it end and start retry from the breaking point.
+   * The value will then update to "undefined", once the injection works.
+   */
+  doInjectErrorOnce?: boolean;
+
+  /**
+   * A threshold, not a limit. Dictates the amount of data that a stream buffers before it stops asking for more data.
+   */
+  highWaterMark?: number;
+}
+
+/**
+ * ONLY AVAILABLE IN NODE.JS RUNTIME.
+ *
+ * A Node.js ReadableStream will internally retry when internal ReadableStream unexpected ends.
+ */
+export class RetriableReadableStream extends Readable {
+  private start: number;
+  private offset: number;
+  private end: number;
+  private getter: ReadableStreamGetter;
+  private source: NodeJS.ReadableStream;
+  private retries: number = 0;
+  private maxRetryRequests: number;
+  private onData?: (data: Buffer) => void;
+  private onEnd?: () => void;
+  private onProgress?: (progress: TransferProgressEvent) => void;
+  private options: RetriableReadableStreamOptions;
+
+  /**
+   * Creates an instance of RetriableReadableStream.
+   *
+   * @param source - The current ReadableStream returned from getter
+   * @param getter - A method calling downloading request returning
+   *                                      a new ReadableStream from specified offset
+   * @param offset - Offset position in original data source to read
+   * @param count - How much data in original data source to read
+   * @param options -
+   */
+  public constructor(
+    source: NodeJS.ReadableStream,
```

Comment:
nitpick: maybe saving some characters with 
```suggestion
    private source: NodeJS.ReadableStream,
```